File: tools/analyze_recent_sales.py
#!/usr/bin/env python3
"""分析 recent_sales CSV，输出半年（2025-10 至 2026-04）聚合 JSON。"""
import csv
import json
from collections import defaultdict
from pathlib import Path

CSV = Path(__file__).resolve().parent.parent.parent / "recent_sales (3).csv"
MONTHS = ["2025-10", "2025-11", "2025-12", "2026-01", "2026-02", "2026-03", "2026-04"]
LABELS = ["10月'25", "11月'25", "12月'25", "1月'26", "2月'26", "3月'26", "4月'26"]
EXCLUDED_PROJECTS = {"Private", "Al Deem Towerhome", "Bal Ghaiylam"}
APT_TYPES = {"apartment"}
VILLA_TYPES = {"villa", "townhouse / attached villa"}


def month_key(date_str: str) -> str:
    return date_str.strip()[:7]


def analyze():
    monthly = {m: {"apt": [], "villa": [], "apt_vol": 0, "villa_vol": 0} for m in MONTHS}
    offplan = {"apt": 0, "villa": 0, "apt_ready": 0, "villa_ready": 0}
    districts_apt = defaultdict(int)
    districts_villa = defaultdict(int)
    districts_all = defaultdict(int)
    sale_type = defaultdict(int)
    total_value = 0.0
    total_count = 0

    with open(CSV, encoding="utf-8-sig") as f:
        reader = csv.reader(f)
        next(reader)
        for row in reader:
            if len(row) < 13 or row[0].strip() != "residential":
                continue
            project = row[8].strip()
            if project in EXCLUDED_PROJECTS:
                continue
            prop = row[1].strip()
            m = month_key(row[2])
            if m not in MONTHS:
                continue
            try:
                rate = float(row[11])
                price = float(row[9])
            except ValueError:
                continue
            if rate <= 0 or rate > 100_000:
                continue

            district = row[6].strip() or "其他"
            app_type = row[12].strip().lower()
            is_offplan = app_type == "off-plan"

            total_count += 1
            total_value += price
            sale_type[row[13].strip() if len(row) > 13 else "unknown"] += 1
            districts_all[district] += 1

            if prop in APT_TYPES:
                monthly[m]["apt_vol"] += 1
                monthly[m]["apt"].append(rate)
                districts_apt[district] += 1
                if is_offplan:
                    offplan["apt"] += 1
                else:
                    offplan["apt_ready"] += 1
            elif prop in VILLA_TYPES:
                monthly[m]["villa_vol"] += 1
                monthly[m]["villa"].append(rate)
                districts_villa[district] += 1
                if is_offplan:
                    offplan["villa"] += 1
                else:
                    offplan["villa_ready"] += 1

    def avg(lst):
        return round(sum(lst) / len(lst)) if lst else 0

    def avg_k(lst):
        return round(avg(lst) / 1000, 1) if lst else 0

    apt_vol = [monthly[m]["apt_vol"] for m in MONTHS]
    villa_vol = [monthly[m]["villa_vol"] for m in MONTHS]
    apt_price_k = [avg_k(monthly[m]["apt"]) for m in MONTHS]
    villa_price_k = [avg_k(monthly[m]["villa"]) for m in MONTHS]

    apt_total = sum(apt_vol)
    villa_total = sum(villa_vol)
    offplan_apt_total = offplan["apt"]
    offplan_villa_total = offplan["villa"]

    # Top districts - merge for apartment chart (top 5 + other)
    def top_n(d: dict, n=5):
        items = sorted(d.items(), key=lambda x: -x[1])
        top = items[:n]
        other = sum(v for _, v in items[n:])
        labels = [k for k, _ in top]
        data = [v for _, v in top]
        if other:
            labels.append("其他区域")
            data.append(other)
        return labels, data

    apt_labels, apt_data = top_n(districts_apt, 5)
    villa_labels, villa_data = top_n(districts_villa, 6)

    # Three islands share for apartments
    three_islands = {"Al Reem Island", "Yas Island", "Al Saadiyat Island"}
    three_vol = sum(districts_apt.get(d, 0) for d in three_islands)
    three_share = round(three_vol / apt_total * 100) if apt_total else 0

    # Price change Oct vs Apr
    apt_oct = avg(monthly["2025-10"]["apt"])
    apt_apr = avg(monthly["2026-04"]["apt"])
    villa_oct = avg(monthly["2025-10"]["villa"])
    villa_apr = avg(monthly["2026-04"]["villa"])

    def pct_change(a, b):
        if not a:
            return 0
        return round((b - a) / a * 100)

    result = {
        "meta": {
            "source": "Abu Dhabi Real Estate Centre 住宅成交记录",
            "periodStart": "2025-10-01",
            "periodEnd": "2026-04-30",
            "periodLabel": "2025年10月 — 2026年4月（近半年）",
            "updated": "2026-06-26",
            "excluded": list(EXCLUDED_PROJECTS),
            "propertyTypes": ["apartment", "villa", "townhouse / attached villa"],
            "csvFile": "recent_sales (3).csv",
            "note": "近半年微观成交；与 ADREC 2025 年报口径不同。",
        },
        "summary": {
            "totalTransactions": total_count,
            "totalValueAedBn": round(total_value / 1e9, 1),
            "peakMonth": MONTHS[apt_vol.index(max(apt_vol + villa_vol))],
            "peakMonthTotal": max(a + b for a, b in zip(apt_vol, villa_vol)),
            "apartmentSharePct": round(apt_total / total_count * 100, 1) if total_count else 0,
            "villaSharePct": round(villa_total / total_count * 100, 1) if total_count else 0,
            "offPlanSharePct": round((offplan_apt_total + offplan_villa_total) / total_count * 100, 1) if total_count else 0,
            "offPlanAptPct": round(offplan_apt_total / apt_total * 100, 1) if apt_total else 0,
            "offPlanVillaPct": round(offplan_villa_total / villa_total * 100, 1) if villa_total else 0,
            "readySharePct": round((offplan["apt_ready"] + offplan["villa_ready"]) / total_count * 100, 1) if total_count else 0,
            "aptPriceChangePct": pct_change(apt_oct, apt_apr),
            "villaPriceChangePct": pct_change(villa_oct, villa_apr),
            "aptPriceStart": apt_oct,
            "aptPriceEnd": apt_apr,
            "villaPriceStart": villa_oct,
            "villaPriceEnd": villa_apr,
            "primarySharePct": round(sale_type.get("primary", 0) / total_count * 100, 1) if total_count else 0,
            "secondarySharePct": round(sale_type.get("secondary", 0) / total_count * 100, 1) if total_count else 0,
        },
        "months": {
            "labels": LABELS,
            "apartmentVolume": apt_vol,
            "villaVolume": villa_vol,
            "apartmentPriceSqmK": apt_price_k,
            "villaPriceSqmK": villa_price_k,
        },
        "structure": {
            "typeSplit": {
                "labels": [f"公寓 {round(apt_total/total_count*100,1)}%", f"别墅/联排 {round(villa_total/total_count*100,1)}%"],
                "data": [apt_total, villa_total],
            },
            "offPlanAll": {
                "labels": [
                    f"期房 {round((offplan_apt_total+offplan_villa_total)/total_count*100,1)}%",
                    f"现房 {round((offplan['apt_ready']+offplan['villa_ready'])/total_count*100,1)}%",
                ],
                "data": [offplan_apt_total + offplan_villa_total, offplan["apt_ready"] + offplan["villa_ready"]],
            },
            "saleSequence": {
                "labels": ["一级市场 primary", "二级市场 secondary"],
                "data": [sale_type.get("primary", 0), sale_type.get("secondary", 0)],
            },
        },
        "regions": {
            "apartment": {
                "labels": apt_labels,
                "data": apt_data,
                "highlight": [i for i, l in enumerate(apt_labels) if l in three_islands],
                "topThreeSharePct": three_share,
            },
            "villa": {"labels": villa_labels, "data": villa_data},
        },
        "highlights": {
            "aptPeakMonth": {"month": MONTHS[apt_vol.index(max(apt_vol))], "volume": max(apt_vol)},
            "villaPeakMonth": {"month": MONTHS[villa_vol.index(max(villa_vol))], "volume": max(villa_vol)},
            "aptPricePeak": {"month": MONTHS[apt_price_k.index(max(apt_price_k))], "priceSqmK": max(apt_price_k)},
            "villaPricePeak": {"month": MONTHS[villa_price_k.index(max(villa_price_k))], "priceSqmK": max(villa_price_k)},
        },
        "adrecCrossRef": {
            "annualOffPlanPct2025": 71,
            "halfYearOffPlanPct": round((offplan_apt_total + offplan_villa_total) / total_count * 100, 1) if total_count else 0,
            "annualCashPct2025": 87,
            "note": "近半年期房占比通常高于或接近年报，因 2025 后段新盘集中入市；一级市场占比高说明仍以开发商 primary 销售为主。",
        },
    }

    out = Path(__file__).resolve().parent.parent / "data" / "halfyear.json"
    out.write_text(json.dumps(result, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    print(json.dumps(result, ensure_ascii=False, indent=2))
    print(f"\nWritten: {out}")


if __name__ == "__main__":
    analyze()

File: tools/analyze_recent_sales.mjs
/**
 * 分析 recent_sales CSV → data/halfyear.json
 * 区间：2025-10 至 2026-04
 * 排除：Private / Al Deem Towerhome / Bal Ghaiylam
 * 输出：capitalRegion（全首都圈匹配记录）+ investmentZone（投资区 District 筛选）
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const CSV = path.join(__dirname, '..', '..', 'recent_sales (3).csv');
const OUT = path.join(__dirname, '..', 'data', 'halfyear.json');
const IZ_CONFIG = path.join(__dirname, '..', 'data', 'investment_zones.json');

const MONTHS = ['2025-10', '2025-11', '2025-12', '2026-01', '2026-02', '2026-03', '2026-04'];
const LABELS = ["10月'25", "11月'25", "12月'25", "1月'26", "2月'26", "3月'26", "4月'26"];
const EXCLUDED_PROJECTS = new Set(['Private', 'Al Deem Towerhome', 'Bal Ghaiylam']);
const APT = new Set(['apartment']);
const VILLA = new Set(['villa', 'townhouse / attached villa']);
const THREE_ISLANDS = new Set(['Al Reem Island', 'Yas Island', 'Al Saadiyat Island']);

function parseCsvLine(line) {
  const out = [];
  let cur = '';
  let inQ = false;
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    if (c === '"') { inQ = !inQ; continue; }
    if (c === ',' && !inQ) { out.push(cur); cur = ''; continue; }
    cur += c;
  }
  out.push(cur);
  return out;
}

function avg(arr) {
  return arr.length ? Math.round(arr.reduce((a, b) => a + b, 0) / arr.length) : 0;
}
function avgK(arr) {
  return arr.length ? Math.round(avg(arr) / 100) / 10 : 0;
}
function avgKOrNull(arr) {
  return arr.length ? Math.round(avg(arr) / 100) / 10 : null;
}
function pctChange(a, b) {
  return a ? Math.round(((b - a) / a) * 100) : 0;
}
function pct(part, whole) {
  return whole ? Math.round((part / whole) * 1000) / 10 : 0;
}

function toYi(aed) {
  return Math.round(aed / 1e7) / 10;
}
function topN(map, n) {
  const items = [...map.entries()].sort((a, b) => b[1] - a[1]);
  const top = items.slice(0, n);
  return { labels: top.map(([k]) => k), data: top.map(([, v]) => v) };
}

function emptyBucket() {
  const monthEntry = () => ({
    apt: [],
    villa: [],
    apt_primary: [],
    villa_primary: [],
    apt_secondary: [],
    villa_secondary: [],
    apt_vol: 0,
    villa_vol: 0,
  });
  return {
    monthly: Object.fromEntries(MONTHS.map(m => [m, monthEntry()])),
    offplan: { apt: 0, villa: 0, apt_ready: 0, villa_ready: 0 },
    districtsApt: new Map(),
    districtsVilla: new Map(),
    saleType: new Map(),
    totalValue: 0,
    totalCount: 0,
  };
}

function addRecord(bucket, row) {
  const m = row.date.slice(0, 7);
  if (!MONTHS.includes(m)) return;
  const { prop, district, rate, price, isOff, seq } = row;
  if (!APT.has(prop) && !VILLA.has(prop)) return;

  bucket.totalCount++;
  bucket.totalValue += price;
  bucket.saleType.set(seq, (bucket.saleType.get(seq) || 0) + 1);

  if (APT.has(prop)) {
    bucket.monthly[m].apt_vol++;
    bucket.monthly[m].apt.push(rate);
    bucket.districtsApt.set(district, (bucket.districtsApt.get(district) || 0) + 1);
    if (isOff) bucket.offplan.apt++; else bucket.offplan.apt_ready++;
    if (seq === 'primary') bucket.monthly[m].apt_primary.push(rate);
    else if (seq === 'secondary') bucket.monthly[m].apt_secondary.push(rate);
  } else {
    bucket.monthly[m].villa_vol++;
    bucket.monthly[m].villa.push(rate);
    bucket.districtsVilla.set(district, (bucket.districtsVilla.get(district) || 0) + 1);
    if (isOff) bucket.offplan.villa++; else bucket.offplan.villa_ready++;
    if (seq === 'primary') bucket.monthly[m].villa_primary.push(rate);
    else if (seq === 'secondary') bucket.monthly[m].villa_secondary.push(rate);
  }
}

function buildSlice(bucket, label) {
  const aptVol = MONTHS.map(m => bucket.monthly[m].apt_vol);
  const villaVol = MONTHS.map(m => bucket.monthly[m].villa_vol);
  const aptPriceK = MONTHS.map(m => avgK(bucket.monthly[m].apt));
  const villaPriceK = MONTHS.map(m => avgK(bucket.monthly[m].villa));
  const aptTotal = aptVol.reduce((a, b) => a + b, 0);
  const villaTotal = villaVol.reduce((a, b) => a + b, 0);
  const total = bucket.totalCount;
  const offTotal = bucket.offplan.apt + bucket.offplan.villa;
  const monthlyTotals = MONTHS.map((m, i) => aptVol[i] + villaVol[i]);
  const peakIdx = monthlyTotals.indexOf(Math.max(...monthlyTotals));

  const threeVol = [...bucket.districtsApt.entries()]
    .filter(([k]) => THREE_ISLANDS.has(k))
    .reduce((s, [, v]) => s + v, 0);

  const aptOct = avg(bucket.monthly['2025-10'].apt);
  const aptApr = avg(bucket.monthly['2026-04'].apt);
  const villaOct = avg(bucket.monthly['2025-10'].villa);
  const villaApr = avg(bucket.monthly['2026-04'].villa);

  const aptRegions = topN(bucket.districtsApt, 6);
  const villaRegions = topN(bucket.districtsVilla, 6);

  return {
    label,
    summary: {
      totalTransactions: total,
      totalValueAedYi: toYi(bucket.totalValue),
      peakMonth: MONTHS[peakIdx],
      peakMonthTotal: monthlyTotals[peakIdx],
      apartmentSharePct: pct(aptTotal, total),
      villaSharePct: pct(villaTotal, total),
      offPlanSharePct: pct(offTotal, total),
      offPlanAptPct: pct(bucket.offplan.apt, aptTotal),
      offPlanVillaPct: pct(bucket.offplan.villa, villaTotal),
      readySharePct: pct(bucket.offplan.apt_ready + bucket.offplan.villa_ready, total),
      aptPriceChangePct: pctChange(aptOct, aptApr),
      villaPriceChangePct: pctChange(villaOct, villaApr),
      aptPriceStart: aptOct,
      aptPriceEnd: aptApr,
      villaPriceStart: villaOct,
      villaPriceEnd: villaApr,
      primarySharePct: pct(bucket.saleType.get('primary') || 0, total),
      secondarySharePct: pct(bucket.saleType.get('secondary') || 0, total),
    },
    months: {
      labels: LABELS,
      apartmentVolume: aptVol,
      villaVolume: villaVol,
      apartmentPriceSqmK: aptPriceK,
      villaPriceSqmK: villaPriceK,
      primary: {
        apartmentPriceSqmK: MONTHS.map(m => avgKOrNull(bucket.monthly[m].apt_primary)),
        villaPriceSqmK: MONTHS.map(m => avgKOrNull(bucket.monthly[m].villa_primary)),
      },
      secondary: {
        apartmentPriceSqmK: MONTHS.map(m => avgKOrNull(bucket.monthly[m].apt_secondary)),
        villaPriceSqmK: MONTHS.map(m => avgKOrNull(bucket.monthly[m].villa_secondary)),
      },
    },
    structure: {
      typeSplit: {
        labels: [`公寓 ${pct(aptTotal, total)}%`, `别墅/联排 ${pct(villaTotal, total)}%`],
        data: [aptTotal, villaTotal],
      },
      offPlanAll: {
        labels: [`期房 ${pct(offTotal, total)}%`, `现房 ${pct(bucket.offplan.apt_ready + bucket.offplan.villa_ready, total)}%`],
        data: [offTotal, bucket.offplan.apt_ready + bucket.offplan.villa_ready],
      },
      saleSequence: {
        labels: ['一级市场', '二级市场'],
        data: [bucket.saleType.get('primary') || 0, bucket.saleType.get('secondary') || 0],
      },
    },
    regions: {
      apartment: {
        labels: aptRegions.labels,
        data: aptRegions.data,
        highlight: aptRegions.labels.map((l, i) => (THREE_ISLANDS.has(l) ? i : -1)).filter(i => i >= 0),
        topThreeSharePct: pct(threeVol, aptTotal),
      },
      villa: { labels: villaRegions.labels, data: villaRegions.data },
    },
    highlights: {
      aptPeakMonth: { month: MONTHS[aptVol.indexOf(Math.max(...aptVol))], volume: Math.max(...aptVol) },
      villaPeakMonth: { month: MONTHS[villaVol.indexOf(Math.max(...villaVol))], volume: Math.max(...villaVol) },
      aptPricePeak: { month: MONTHS[aptPriceK.indexOf(Math.max(...aptPriceK))], priceSqmK: Math.max(...aptPriceK) },
      villaPricePeak: { month: MONTHS[villaPriceK.indexOf(Math.max(...villaPriceK))], priceSqmK: Math.max(...villaPriceK) },
    },
  };
}

// ── 读取 CSV ──
const izConfig = JSON.parse(fs.readFileSync(IZ_CONFIG, 'utf8'));
const IZ = new Set(izConfig.districts);

const allBucket = emptyBucket();
const izBucket = emptyBucket();

const text = fs.readFileSync(CSV, 'utf8');
for (const line of text.split(/\r?\n/).slice(1)) {
  if (!line.trim()) continue;
  const row = parseCsvLine(line);
  if (row.length < 13 || row[0].trim() !== 'residential') continue;
  if (EXCLUDED_PROJECTS.has(row[8].trim())) continue;

  const rate = parseFloat(row[11]);
  const price = parseFloat(row[9]);
  if (!(rate > 0 && rate <= 100000)) continue;

  const rec = {
    date: row[2].trim(),
    prop: row[1].trim(),
    district: row[6].trim() || '其他',
    rate,
    price,
    isOff: row[12].trim().toLowerCase() === 'off-plan',
    seq: (row[13] || 'unknown').trim(),
  };

  addRecord(allBucket, rec);
  if (IZ.has(rec.district)) addRecord(izBucket, rec);
}

const all = buildSlice(allBucket, '首都圈成交（已排除三类项目）');
const investmentZone = buildSlice(izBucket, '永久产权投资区');

const result = {
  meta: {
    source: 'Abu Dhabi Real Estate Centre 住宅成交记录',
    periodStart: '2025-10-01',
    periodEnd: '2026-04-30',
    periodLabel: '2025年10月 — 2026年4月（近半年）',
    updated: '2026-06-26',
    excludedProjects: [...EXCLUDED_PROJECTS],
    propertyTypes: ['apartment', 'villa', 'townhouse / attached villa'],
    csvFile: 'recent_sales (3).csv',
    investmentZoneDistricts: izConfig.districts.length,
    note: '已排除 Private / Al Deem Towerhome / Bal Ghaiylam 三类项目；投资区按 District 字段匹配 ADREC 名录。',
  },
  all,
  investmentZone,
  investmentZoneMeta: izConfig.meta,
  compare: {
    izShareOfTransactionsPct: pct(investmentZone.summary.totalTransactions, all.summary.totalTransactions),
    izShareOfValuePct: pct(investmentZone.summary.totalValueAedYi, all.summary.totalValueAedYi),
    nonIzTransactions: all.summary.totalTransactions - investmentZone.summary.totalTransactions,
    topNonIzDistricts: [...[...allBucket.districtsApt.entries(), ...allBucket.districtsVilla.entries()]
      .reduce((m, [k, v]) => { if (!IZ.has(k)) m.set(k, (m.get(k) || 0) + v); return m; }, new Map())
      .entries()]
      .sort((a, b) => b[1] - a[1])
      .slice(0, 5)
      .map(([district, count]) => ({ district, count })),
  },
  adrecCrossRef: {
    annualOffPlanPct2025: 71,
    halfYearOffPlanPctAll: all.summary.offPlanSharePct,
    halfYearOffPlanPctIz: investmentZone.summary.offPlanSharePct,
    annualCashPct2025: 87,
    annualTransactionValueAedYi: 761,
    halfYearValueAllAedYi: all.summary.totalValueAedYi,
    halfYearValueIzAedYi: investmentZone.summary.totalValueAedYi,
    stockShareInvestmentZonePct: 21,
    note: '投资区成交占近半年总量约七成——与 ADREC「投资区存量约占 21%」不矛盾：小盘子里交易更活跃；购前须 DARI 核验单元产权。',
  },
};

fs.writeFileSync(OUT, JSON.stringify(result, null, 2) + '\n', 'utf8');
console.log('All:', all.summary.totalTransactions, '| IZ:', investmentZone.summary.totalTransactions, `(${result.compare.izShareOfTransactionsPct}%)`);
console.log('Written:', OUT);
